Feature: A setting to bring back old tile tooltip behavior.

RoqueDeicide opened the pull request:
If enabled, the setting allows tooltips for tiles, such as industries, to show instantly, when otherwise they would require a right-click.

## Motivation / Problem

One of the features in this patchpack is showing tooltips when hovering over tiles on the map. In ancient days it used to be that , if "Show tooltips" setting was set to "Right-click", a right-click would only be required to show a tooltip for an interface widget, while hovering over, for example, an industry would show a tooltip instantly. This was fixed in v0.47.1 release, creating 2 problems:

1. There is a difference between utility of interface and town/industry tooltips for the player. The former provides information that, generally, only changes with updates to the program, while the latter provides real-time gameplay information that may frequently change throughout the game. This means that there is a good reason for the player to want to have widget tooltips to be right-click-only, so they don't create visual clutter with info that is already known, while allowing instant access to latest updates to industry production rates. The above-mentioned change prevents this.
2. Requiring a right-click on a tile to see a tooltip conflicts with how OpenTTD handles scrolling. In order for a tooltip to be shown, a "OnMouseOver" function must be called, but that doesn't happen if a player is moving the map or a viewport. This means, that in order for a tooltip to show for a tile with a right-click, the scrolling must be bound to the left mouse button, which is 1 scrolling mode out of 4. In other words, in most scrolling modes, setting tooltips to show on a right-click, effectively disables town/industry ones.

## Description

This adds an advanced setting to instantly show tile tooltips, if "Show tooltips" setting is set to "Right-click". It's located in "Interface/General" and is called "Show tooltips for tiles without a right-click". It's enabled by default to avoid the most likely situation where a "Right-click" setting confuses the player into believing that town/industry tooltips are incompatible with it.

## Limitations

Some things I've noted when implementing this feature:
* There appears to be code for station rating tooltips, but it doesn't appear to work for me. I've updated the if condition for it regardless.
* There could be 2 separate settings for widget and tile tooltips, but I chose to not go that far. Same with allowing "MouseOver" events to during scrolling.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

JGRennison commented:
There is a setting for station rating tooltips, however it is enabled by default.
I don't see that it is necessary to make any changes to this tooltip as it is not viewport/tile-based. (See the ratings section in station view windows).

I have no objections to this change in principle, but I don't really agree that it's necessary to enable it by default.
The motivation for fixing the tooltip behaviour to match the user-specified behaviour was that it didn't match the expectations of that subset of users who used right-click tooltips at the time.

The change to the if statement does not match the previous behaviour when this setting is off.

Ideally the new language strings should go in src/lang/extra/, but that's something I can tidy up on merge.

RoqueDeicide commented:
> There is a setting for station rating tooltips, however it is enabled by default. I don't see that it is necessary to make any changes to this tooltip as it is not viewport/tile-based. (See the ratings section in station view windows).

Sorry about that, I got confused. I've removed the change.

> I have no objections to this change in principle, but I don't really agree that it's necessary to enable it by default. The motivation for fixing the tooltip behaviour to match the user-specified behaviour was that it didn't match the expectations of that subset of users who used right-click tooltips at the time.

After some thought, I'm not sure if making it on by default matters that match, considering that "show tooltip on right-click" is not a default setting, so if it's changed, then the player will probably notice the setting right below it and make a decision on what they want right there. People who are migrating from variants of OpenTTD, however, may have "right-click" option and without this setting on by default, they may not even realize that town/industry tooltips are a feature, so I'm leaning towards 
it being set to on by default, but it can be changed on merge.

> The change to the if statement does not match the previous behaviour when this setting is off.

Can you specify? I don't see any changes in the game itself, the code is slightly different, but, I think, it should arrive at the same destination.
 
> Ideally the new language strings should go in src/lang/extra/, but that's something I can tidy up on merge.

I've moved the strings to extra/english.txt

JGRennison commented:
Thanks I've merged this with some minor adjustments
